File: 141-linked-list-cycle/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        hashmap = {head.val:0}
        
        while head.next != None:
            next_node = head.next
            if next_node in hashmap:
                return True

            hashmap[next_node] = 0
            head = next_node
        
        return False
            
        

File: 141-linked-list-cycle/linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr) {
            return false;
        }

        set<ListNode*> traversed_nodes;
        traversed_nodes.insert(head);

        while( head->next != nullptr) {
            if (traversed_nodes.count(head->next) != 0) {
                return true;
            }

            traversed_nodes.insert(head);
            head = head->next;
        }

        return false;
        
    }
};